### PR TITLE
Account abstraction EIP

### PR DESCRIPTION
**Pull request by vbuterin**
Account abstraction (AA) allows a contract to be the top-level account that pays fees and starts transaction execution.

### CODE REVIEW

**Review by lightclient — APPROVED** [review]

File: EIPS/eip-2938.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+eip: 2938
+title: Account Abstraction
+author: Vitalik Buterin (@vbuterin), Ansgar Dietrichs (@adietrichs), Matt Garnett (@lightclient), Will Villanueva (@villanuevawill), Sam Wilson (@SamWilsn)
+discussions-to: https://ethereum-magicians.org/t/implementing-account-abstraction-as-part-of-eth1-x/4020
+status: Draft
+type: Standards Track
+category: Core
+created: 2020-09-04
+requires: 1344, 2718
+replaces: 86
```
Comment (MicahZoltu) [inline EIPS/eip-2938.md:11]:
```suggestion
```
You cannot replace/supersede a draft transaction, only a final one.

Note: just saw this at a quick glance, will do a full review later.

File: EIPS/eip-2938.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+Account abstraction allows for the validity conditions of a transaction to include execution of arbitrary EVM code (with some limits regarding what state can be accessed), allowing the above use cases.
+
+We split account abstraction into two tiers: **single-tenant AA**, which is intended to support abstracted accounts that represent wallets or otherwise are only meant to be accessed by one or a few participants, and **multi-tenant AA**, which is intended to allow abstracted accounts to also directly be applications (eg. tornado.cash, Uniswap). Out of the four use cases above, it can be approximately said that single-tenant AA supports (1) and (2), and multi-tenant AA supports (3) and (4). We get into the differences between the two tiers in the specification and rationale section below.
+
+## Specification (Single-tenant AA)
```
Comment (MicahZoltu) [inline EIPS/eip-2938.md:35]:
```suggestion
## Specification
### Single-tenant AA
```
We prefer the sections align with the template, as this allows automated tools and rendering engines to do fancy stuff in the future.

File: EIPS/eip-2938.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+## Test Cases
+See: https://github.com/quilt/tests/tree/account-abstraction
+
+## Implementation
+See: https://github.com/quilt/go-ethereum/tree/account-abstraction
+
+## Security Considerations
+
+See https://ethresear.ch/t/dos-vectors-in-account-abstraction-aa-or-validation-generalization-a-case-study-in-geth/7937 for an analysis of DoS issues.
```
Comment (MicahZoltu) [inline EIPS/eip-2938.md:246]: At some point before final security considerations should be inlined into this document, but that isn't necessary for draft status.

File: EIPS/eip-2938.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+## Backwards Compatibility
+
+This AA implementation preserves the existing transaction type. The use of `assert origin == caller` to verify that an account is an EOA remains sound, but is not extensible to AA accounts; AA transactions will always have `origin == AA_ENTRY_POINT`.
+
+Badly-designed single-tenant AA contracts will break the transaction non-malleability invariant. That is, it is possible to take an AA transaction in-flight, modify it, and have the modified version still be valid; AA account contracts can be designed in such a way as to make that not possible, but it is their responsibility. Multi-tenant AA will break the transaction non-malleability invariant much more thoroughly, making the transaction hash unpredictable even for legitimate applications that use the multi-tenant AA features (though the invariant will not further break for applications that existed before then).
+
+All AA contracts may also potentially not have replay protection, unless they build it in explicitly; this can be done with the `CHAINID (0x46)` opcode introduced in [EIP 1344](./eip-1344.md).
+
+## Test Cases
+See: https://github.com/quilt/tests/tree/account-abstraction
```
Comment (MicahZoltu) [inline EIPS/eip-2938.md:239]: Ideally, test cases will be inlined into this document, or added as an asset to `../assets/eip-2938/whatever.wat`, but that isn't necessary for draft status.

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
Only the Abstract inclusion suggestion and the Specification section naming suggestion are required for merging as a draft.  The rest is just generalized feedback that you can ignore until Last Call.

File: EIPS/eip-2938.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+Account abstraction allows for the validity conditions of a transaction to include execution of arbitrary EVM code (with some limits regarding what state can be accessed), allowing the above use cases.
+
+We split account abstraction into two tiers: **single-tenant AA**, which is intended to support abstracted accounts that represent wallets or otherwise are only meant to be accessed by one or a few participants, and **multi-tenant AA**, which is intended to allow abstracted accounts to also directly be applications (eg. tornado.cash, Uniswap). Out of the four use cases above, it can be approximately said that single-tenant AA supports (1) and (2), and multi-tenant AA supports (3) and (4). We get into the differences between the two tiers in the specification and rationale section below.
+
+## Specification (Single-tenant AA)
+
+After `FORK_BLOCK`, the following changes will be recognized by the protocol.
```
Comment (holgerd77) [inline EIPS/eip-2938.md:37]: This would be "Starting with `FORK_BLOCK`" I guess?

File: EIPS/eip-2938.md [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+| Constant | Value |
+| - | - |
+| **`AA_ENTRY_POINT`** | `0xffffffffffffffffffffffffffffffffffffffff` |
+| **`AA_TX_TYPE`** | `2` |
+| **`FORK_BLOCK`** | TBD |
+| **`AA_BASE_GAS_COST`** | 15000 |
+
+### New Transaction Type
+
+A new EIP-2718 transaction with type `AA_TX_TYPE` is introduced. Transactions of this type are referred to as "AA transactions". Their payload should be interpreted as `rlp([nonce, target, data])`.
```
Comment (holgerd77) [inline EIPS/eip-2938.md:50]: Can the "EIP-2718" mention be supplemented by the EIP title (in brackets) here for readability?

Comment (MicahZoltu) [inline EIPS/eip-2938.md:50]:
```suggestion
A new [EIP-2718](./eip-2718.md) transaction with type `AA_TX_TYPE` is introduced. Transactions of this type are referred to as "AA transactions". Their payload should be interpreted as `rlp([nonce, target, data])`.
```

File: EIPS/eip-2938.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+
+## Abstract
+
+**See also: https://ethereum-magicians.org/t/implementing-account-abstraction-as-part-of-eth1-x/4020 and the links therein for historical work and motivation.**
+
+Transaction validity, as of Muir Glacier, is defined rigidly by the protocol: ECDSA signature, a simple nonce, and account balance. Account abstraction extends the validity conditions of transactions with the execution of arbitrary EVM bytecode (with some limits on what state may be accessed.) To signal validity, we propose a new EVM opcode `PAYGAS`, which also sets the gas price and gas limit the contract is willing to pay.
+
+We split account abstraction into two tiers: **single-tenant AA**, which is intended to support wallets or other use cases with few participants, and **multi-tenant AA**, which is intended to support applications with many participants (eg. tornado.cash, Uniswap.)
+
+### Motivation
```
Comment (MicahZoltu) [inline EIPS/eip-2938.md:25]:
```suggestion
## Motivation
```

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
Missed one required change in my previous review.

File: EIPS/eip-2938.md [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+
+| Variable | Type | Initial value |
+| - | - | - |
+| `globals.transaction_fee_paid` | `bool` | `False if type(tx) == AA_TX_TYPE else True` |
+| `globals.gas_price` | `int` | `0 if type(tx) == AA_TX_TYPE else tx.gas_price` |
+| `globals.gas_limit` | `int` | `0 if type(tx) == AA_TX_TYPE else tx.gas_limit` |
+
+#### `NONCE (0xAB)` Opcode
+
+A new opcode `NONCE (0xAB)` is introduced, with gas cost `G_base`, which pushes the `tx.nonce` onto the stack.
```
Comment (axic) [inline EIPS/eip-2938.md:76]:
```suggestion
A new opcode `NONCE (0xAB)` is introduced, with gas cost `G_base`, which pushes the `account.nonce` onto the stack.
```

If `tx.nonce` would be used, I would suggest to name this `TXNONCE`. Luckily however `tx.nonce == account.nonce`, so changing the description is enough.

File: EIPS/eip-2938.md [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+#### `PAYGAS (0xAA)` Opcode
```
Comment (axic) [inline EIPS/eip-2938.md:78]: While I see it is clever to use `0xAA` and `0xAB` as opcodes, I think if this is adopted these opcodes should fit into appropriate sections: which is likely `0x48` and` 0x49`.

As a draft the current proposed ones are fine, perhaps I should leave this comment on the discussion url.

File: EIPS/eip-2938.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+
+A new opcode `PAYGAS (0xAA)` is introduced, with gas cost `G_base`. It takes two arguments off the stack: (top) `version_number`, (second from top) `memory_start`. In the initial implementation, it will `assert version_number == 0` and read:
```
Comment (axic) [inline EIPS/eip-2938.md:80]: Curious: why a version and what future versions are envisioned?

File: EIPS/eip-2938.md (hunk 4 above)
Comment (axic) [reply EIPS/eip-2938.md:245]: `.wat`, are you suggesting the introduction of WebAssembly? 🙊

File: EIPS/eip-2938.md (hunk 10 above)
Comment (vbuterin) [reply EIPS/eip-2938.md:80]: * EIP 1559
* Escalator
* Block hash dependencies

**Review by MicahZoltu — APPROVED** [review]
I recommend reviewing the feedback that has been left, but that can be applied after merging as draft.

**Review by KaycoinsInc — APPROVED** [review]

**Review by KaycoinsInc — APPROVED** [review]

**Pull request merged**